N°5090 Setup : improve missing dependencies message

piRGoif opened the pull request:
To reproduce, simply add a single module that is part of a multi module extension. 
For example itop-approval-portal or itop-standard-email-synchro

Before PR you'll get : 

![image](https://user-images.githubusercontent.com/8947448/159303923-3e5699d0-f4b6-4d9a-8514-5d9ec42e5f06.png)

After PR : 

![image](https://user-images.githubusercontent.com/8947448/159687942-22d70ff6-f17c-44d8-a5d8-65c5c6bebe94.png)

I also changed the MissingDependencyException parent (from Exception to CoreException), as this is an iTop exception !

piRGoif committed:
Change dependencies check message format : one faulty message per line

piRGoif committed:
More formatting : each missing dependency on a line

piRGoif committed:
Fix PHPDoc

Molkobain commented on setup/modulediscovery.class.inc.php:
```diff
@@ -19,9 +19,42 @@
+	 */
 	public $aModulesInfo;
+
+	public function GetMessageForHtml()
```
Some PHPDoc and `@since` could be cool

piRGoif replied:
Done in a1cc8d1c

Molkobain commented on setup/modulediscovery.class.inc.php:
```diff
@@ -19,9 +19,42 @@
+			$sModuleLabel = $aModuleErrors['module']['label'];
+			$aModuleMissingDependencies = $aModuleErrors['dependencies'];
+			$sErrorMessage .= <<<HTML
+	<li><strong>{$sModuleLabel} ({$sModuleId})</strong>:
```
I find it a bit hard to read having the impacted module name in bold. It seems to big to me and I would rather have it all in regular weight, the hierarchy is enough to understand, but not a blocking point to me.

Hipska replied:
Maybe only the label in strong, and the module Id normal?

Molkobain replied:
Worth a try

piRGoif replied:
Without bold : 

![image](https://user-images.githubusercontent.com/8947448/159687169-d3dde839-74a2-447e-8980-7cae56757dbc.png)

Bold only on module label only: 

![image](https://user-images.githubusercontent.com/8947448/159687375-42595af9-7274-48dd-9910-7be4b3899be8.png)

And module code only:
![image](https://user-images.githubusercontent.com/8947448/159687501-3d94cca3-0dc9-4ad7-887e-f89e596c373d.png)

I like the second option ! 
Will commit that

piRGoif replied:
Done in 32896df9

Hipska replied:
I do like it as well, great!

Molkobain replied:
Great indeed 🙌

Molkobain commented:
Very nice initiative, should have been done so long ago... 😅😅😅

piRGoif committed:
Update print layout
All the module line were in bold, now only the module label

piRGoif committed:
:bulb: PHPDoc for \MissingDependencyException::GetMessageForHtml

jbostoen commented:
Looks much nicer!

Is this also tackled in the CLI (unattended install)?

piRGoif commented:
> Looks much nicer!

Nice to see people liking this change :)
 
> Is this also tackled in the CLI (unattended install)?

I'll have to take a look !
Actually I didn't change the original behavior : the exception message is still set as before, so other consumers calling $e->getMessage() will still see the same thing.

I'll look at the code and keep you posted !

piRGoif commented:
> Is this also tackled in the CLI (unattended install)?

The [unattended_install script](https://www.itophub.io/wiki/page?id=latest:advancedtopics:automatic_install) will call this stack :

```
\SetupUtils::CheckSelectedModules
   \ModuleDiscovery::GetAvailableModules
      \ModuleDiscovery::GetModules
         \ModuleDiscovery::OrderModulesByDependencies
```

The MissingDependencyException will be thrown in \ModuleDiscovery::OrderModulesByDependencies
It will be catched in \SetupUtils::CheckSelectedModules and the exception's message will be thrown in the returned $aResult array.

So, no change here.

Molkobain commented on setup/modulediscovery.class.inc.php:
```diff
@@ -19,9 +19,46 @@
+	 * @return string HTML to print to the user the modules impacted
+	 * @since 2.7.7 3.0.2 3.1.0 PR #280
+	 */
+	public function GetMessageForHtml()
```
Reviewed today, @piRGoif will rename the method to match `CoreException::getHtmlDesc()`

piRGoif replied:
Done in b4840cca

piRGoif committed:
Rename exception method to the existing getHtmlDesc()
Was asked during the technical PR review to keep consistency between CoreException children

Molkobain commented:
Accepted in technical review.

Molkobain commented:
Accepted in functional review, will be part of 2.7.7.
@piRGoif can you create the corresponding ticket and update the PR's title before we merge?

piRGoif commented:
Done !
Merging...
